examples_tests/61.BlitFilterTest/normalization/normalization.glsl

```diff
@@ -0,0 +1,92 @@
+#ifndef _NBL_GLSL_BLIT_NORMALIZATION_INCLUDED_
+#define _NBL_GLSL_BLIT_NORMALIZATION_INCLUDED_
+
+#include <nbl/builtin/glsl/workgroup/arithmetic.glsl>
+
+uint integerDivide_64_32_32(in uint dividendMsb, in uint dividendLsb, in uint divisor)
+{
+	const uint MAX_UINT = ~0u;
+	const float msbRatio = float(dividendMsb) / float(divisor);
+	const uint quotient = uint((msbRatio * MAX_UINT) + msbRatio) + dividendLsb / divisor;
+	return quotient;
+}
+
+#ifndef _NBL_GLSL_BLIT_NORMALIZATION_MAIN_DEFINED_
+
+#ifndef _NBL_GLSL_SCRATCH_SHARED_DEFINED_
+#error "_NBL_GLSL_SCRATCH_SHARED_DEFINED_ must be defined"
+#endif
+
+#define scratchShared _NBL_GLSL_SCRATCH_SHARED_DEFINED_
+
+#include <../normalization/parameters.glsl>
+nbl_glsl_blit_normalization_parameters_t nbl_glsl_blit_normalization_getParameters();
+
+nbl_glsl_blit_normalization_pixel_t nbl_glsl_blit_normalization_getPaddedData(in ivec3 coord);
+void nbl_glsl_blit_normalization_setData(in nbl_glsl_blit_normalization_pixel_t data, in ivec3 coord);
+uint nbl_glsl_blit_normalization_getAlphaHistogramData(in uint index);
+uint nbl_glsl_blit_normalization_getPassedInputPixelCountData();
+
+void nbl_glsl_blit_normalization_main()
+{
+	// Todo(achal): Assert on the CPU that _NBL_GLSL_BLIT_NORMALIZATION_BIN_COUNT_ == WORKGROUP_SIZE
+	uint histogramVal = 0u;
+	if (gl_LocalInvocationIndex < _NBL_GLSL_BLIT_NORMALIZATION_BIN_COUNT_)
+		histogramVal = nbl_glsl_blit_normalization_getAlphaHistogramData(gl_LocalInvocationIndex);
+
+	const uint cumHistogramVal = nbl_glsl_workgroupInclusiveAdd(histogramVal);
+
+	if (gl_LocalInvocationIndex < _NBL_GLSL_BLIT_NORMALIZATION_BIN_COUNT_)
+		scratchShared[gl_LocalInvocationIndex] = cumHistogramVal;
+	barrier();
+
+	const nbl_glsl_blit_normalization_parameters_t params = nbl_glsl_blit_normalization_getParameters();
+
+	const uint outputPixelCount = params.outImageDim.x * params.outImageDim.y * params.outImageDim.z;
+
+	uint productMsb, productLsb;
+	umulExtended(nbl_glsl_blit_normalization_getPassedInputPixelCountData(), outputPixelCount, productMsb, productLsb);
+
+	const uint pixelsShouldPassCount = integerDivide_64_32_32(productMsb, productLsb, params.inPixelCount);
+	const uint pixelsShouldFailCount = outputPixelCount - pixelsShouldPassCount;
+
+	uint bucketIndex;
+	{
+		uint begin = 0u;
+		const uint end = _NBL_GLSL_BLIT_NORMALIZATION_BIN_COUNT_;
+		const uint value = pixelsShouldFailCount;
+		uint len = end - begin;
+		if (NBL_GLSL_IS_NOT_POT(len))
+		{
+			const uint newLen = 0x1u << findMSB(len);
+			const uint diff = len - newLen;
+
+			begin = NBL_GLSL_LESS(value, NBL_GLSL_EVAL(scratchShared)[newLen]) ? 0u : diff;
+			len = newLen;
+		}
+
+		while (len != 0u)
+		{
+			begin += NBL_GLSL_LESS(value, NBL_GLSL_EVAL(scratchShared)[begin + (len >>= 1u)]) ? 0u : len;
+			begin += NBL_GLSL_LESS(value, NBL_GLSL_EVAL(scratchShared)[begin + (len >>= 1u)]) ? 0u : len;
+		}
+
+		bucketIndex = begin + (NBL_GLSL_LESS(value, NBL_GLSL_EVAL(scratchShared)[begin]) ? 0u : 1u);
+	}
```

Comment:
ugh we already have a `lower_bound` and `upper_bound` implementation!

Reply:
Yeah that's where I copied this from.

In the above, I've used `upper_bound` which I wasn't able to directly use from the `algorithms.glsl` header because it has a slight copypasta bug [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a16fec199ca775bfc41a0f8e03aff90ff37197bb/include/nbl/builtin/glsl/algorithm.glsl#L9).

That said, I now think using `upper_bound` was a mistake because it will _always_ fail more than `pixelsShouldFailCount` even if the the cumulative histogram has an element equal to `pixelsShouldFailedCount`, whereas `lower_bound` will fail exactly `pixelsShouldFailCount` if the cumulative histogram has it, otherwise it would behave like `upper_bound` and fail more pixels than required.

So now I'll be changing this to use `lower_bound`.

Reply:
Nope even `lower_bound` doesn't seem to work.

Gives the compiler error:
>  error: 'lower_bound_scratchShared_NBL_GLSL_LESS' : no matching overloaded function found

Check [this commit](https://github.com/Devsh-Graphics-Programming/Nabla/pull/343/commits/4d20050ab7e9243a157373d471a5f53704440e2d), I believe its the correct usage.

Reply:
I also think the function will have a different name than `lower_bound_scratchShared_NBL_GLSL_LESS`

Why?
Because `scratchShared` is a `#define` in your code, and my function declaration macro will resolve it while concatenating.

So maybe your `NBL_GLSL_CALL_LOWER_BOUND` macro idea is good.

Reply:
> Because scratchShared is a #define in your code, and my function declaration macro will resolve it while concatenating.

Yeah thats what I thought, so I tried with our standard shared array name `_NBL_GLSL_SCRATCH_SHARED_DEFINED_`, but then I remembered thats another  `#define` xD

Reply:
you're looking for a uniform value (same for every invocation in the dispatch) for an image with no layers

and for dynamically uniform value (same for every invocation in the workgroup) for an image with layers

see discord for how to search better, no binary search necessary, just a workgroup inovcation election and value broadcast.

Reply:
Awesome! This is exactly right.

Reply:
Resolved.